From ad25296ffdf79798d46f303bd5ca217f940f9f31 Mon Sep 17 00:00:00 2001
From: Nakata Maho <maho.nakata@gmail.com>
Date: Mon, 28 Sep 2026 10:55:23 +0000
Subject: [PATCH] Add FlagProfile::Upd9002 for the NEC uPD9002 Z80 emulation
 mode

Per-instance flag profile with two values:

  Upd780   default; documented and undocumented Zilog Z80 behaviour as
           in the NEC uPD780C second source (existing behaviour)
  Upd9002  NEC uPD9002 Z80 emulation mode (PC-88VA V1/V2 mode)

Upd9002 applies the rules reproduced from ZEXDOC/ZEXALL on a real
PC-88VA2 in V2 mode:

  R1  F bits 5/3 are never set; POP AF, setAF/setAF2 and reset mask them
  R2  AND: H = 0
  R3  BIT n,r / (HL) / (IX+d) / (IY+d): H = 0
  R4  RLCA/RRCA/RLA/RRA: H and N preserved
  R5  LDI/LDD/LDIR/LDDR: H and N preserved
  R6  ADD HL/IX/IY,rr: H preserved
  R7  ADC/SBC HL,rr: H = carry/borrow out of bit 3

DAA/CPL/SCF/CCF are unresolved and keep Upd780 behaviour.
---
 z80.hpp | 67 +++++++++++++++++++++++++++++++++++++++++++++------------
 1 file changed, 53 insertions(+), 14 deletions(-)

diff --git a/z80.hpp b/z80.hpp
index ac69bb9..a7dfd8f 100644
--- a/z80.hpp
+++ b/z80.hpp
@@ -116,15 +116,40 @@ class Z80
         consumeClock(clock);
     }
 
+  public: // Flag profile
+    // Upd780: documented and undocumented Zilog Z80 flag behaviour, as in the
+    // NEC uPD780C second source. This is the default.
+    // Upd9002: NEC uPD9002 Z80 emulation mode (PC-88VA V1/V2 mode). Rules
+    // R1-R7 are reproduced from ZEXDOC/ZEXALL runs on a real PC-88VA2;
+    // DAA/CPL/SCF/CCF are unresolved and keep Upd780 behaviour.
+    enum class FlagProfile { Upd780, Upd9002 };
+    void setFlagProfile(FlagProfile profile)
+    {
+        flagProfile = profile;
+        maskUndocumentedFlags();
+    }
+    FlagProfile getFlagProfile() const { return flagProfile; }
+
   private: // Internal functions & variables
+    FlagProfile flagProfile = FlagProfile::Upd780;
+    inline bool isUpd9002() const { return flagProfile == FlagProfile::Upd9002; }
+    // R1 (storage variant): F bits 5/3 are never held. Whether POP AF can
+    // store them on real hardware is untested (probe P7).
+    inline void maskUndocumentedFlags()
+    {
+        if (isUpd9002()) {
+            reg.pair.F &= 0xD7;
+            reg.back.F &= 0xD7;
+        }
+    }
     // bit table
     const unsigned char bits[8] = {0b00000001, 0b00000010, 0b00000100, 0b00001000, 0b00010000, 0b00100000, 0b01000000, 0b10000000};
     // flag setter
     inline void setFlagS() { reg.pair.F |= flagS(); }
     inline void setFlagZ() { reg.pair.F |= flagZ(); }
-    inline void setFlagY() { reg.pair.F |= flagY(); }
+    inline void setFlagY() { if (!isUpd9002()) reg.pair.F |= flagY(); } // R1
     inline void setFlagH() { reg.pair.F |= flagH(); }
-    inline void setFlagX() { reg.pair.F |= flagX(); }
+    inline void setFlagX() { if (!isUpd9002()) reg.pair.F |= flagX(); } // R1
     inline void setFlagPV() { reg.pair.F |= flagPV(); }
     inline void setFlagN() { reg.pair.F |= flagN(); }
     inline void setFlagC() { reg.pair.F |= flagC(); }
@@ -418,8 +443,16 @@ class Z80
     inline unsigned short getHL() { return make16BitsFromLE(reg.pair.L, reg.pair.H); }
     inline unsigned short getHL2() { return make16BitsFromLE(reg.back.L, reg.back.H); }
 
-    inline void setAF(unsigned short value) { splitTo8BitsPair(value, &reg.pair.A, &reg.pair.F); }
-    inline void setAF2(unsigned short value) { splitTo8BitsPair(value, &reg.back.A, &reg.back.F); }
+    inline void setAF(unsigned short value)
+    {
+        splitTo8BitsPair(value, &reg.pair.A, &reg.pair.F);
+        maskUndocumentedFlags(); // R1
+    }
+    inline void setAF2(unsigned short value)
+    {
+        splitTo8BitsPair(value, &reg.back.A, &reg.back.F);
+        maskUndocumentedFlags(); // R1
+    }
     inline void setBC(unsigned short value) { splitTo8BitsPair(value, &reg.pair.B, &reg.pair.C); }
     inline void setBC2(unsigned short value) { splitTo8BitsPair(value, &reg.back.B, &reg.back.C); }
     inline void setDE(unsigned short value) { splitTo8BitsPair(value, &reg.pair.D, &reg.pair.E); }
@@ -923,6 +956,7 @@ class Z80
     {
         ctx->reg.pair.F = ctx->pop(3);
         ctx->reg.pair.A = ctx->pop(3);
+        ctx->maskUndocumentedFlags(); // R1
 #ifndef Z80_DISABLE_DEBUG
         if (ctx->isDebug()) ctx->log("[%04X] POP AF <SP:$%04X> = $%04X", ctx->reg.PC - 1, ctx->reg.SP - 2, ctx->getAF());
 #endif
@@ -1836,9 +1870,9 @@ class Z80
         setBC(bc);
         setDE(de);
         setHL(hl);
-        resetFlagH();
+        if (!isUpd9002()) resetFlagH(); // R5
         setFlagPV(bc != 0);
-        resetFlagN();
+        if (!isUpd9002()) resetFlagN(); // R5
         unsigned char an = reg.pair.A + n;
         setFlagY(an & 0b00000010);
         setFlagX(an & 0b00001000);
@@ -2022,8 +2056,10 @@ class Z80
     inline void setFlagByRotate(unsigned char n, bool carry, bool isA = false)
     {
         setFlagC(carry);
-        resetFlagH();
-        resetFlagN();
+        if (!(isA && isUpd9002())) { // R4
+            resetFlagH();
+            resetFlagN();
+        }
         setFlagXY(n);
         if (!isA) {
             setFlagS(n & 0x80);
@@ -3711,7 +3747,7 @@ class Z80
         resetFlagN();
         setFlagXY((result & 0xFF00) >> 8);
         setFlagC((carrybits & 0x10000) != 0);
-        setFlagH((carrybits & 0x1000) != 0);
+        if (!isUpd9002()) setFlagH((carrybits & 0x1000) != 0); // R6
     }
 
     inline void setFlagByAdc16(unsigned short before, unsigned short addition, unsigned char carry)
@@ -3763,6 +3799,7 @@ class Z80
         unsigned char c = isFlagC() ? 1 : 0;
         reg.WZ = hl + 1;
         setFlagByAdc16(hl, nn, c);
+        if (isUpd9002()) setFlagH(((hl & 0xF) + (nn & 0xF) + c) > 0xF); // R7
         setHL(hl + c + nn);
         consumeClock(7);
     }
@@ -3900,6 +3937,7 @@ class Z80
         unsigned char c = isFlagC() ? 1 : 0;
         reg.WZ = hl + 1;
         setFlagBySbc16(hl, nn, c);
+        if (isUpd9002()) setFlagH((hl & 0xF) < ((nn & 0xF) + c)); // R7
         setHL(hl - c - nn);
         consumeClock(7);
     }
@@ -3918,7 +3956,7 @@ class Z80
     {
         reg.pair.A &= n;
         setFlagByLogical();
-        setFlagH();
+        setFlagH(!isUpd9002()); // R2
     }
 
     inline void or8(unsigned char n)
@@ -4373,7 +4411,7 @@ class Z80
         setFlagZ(n ? false : true);
         setFlagPV(isFlagZ());
         setFlagS(!isFlagZ() && 7 == bit);
-        setFlagH();
+        setFlagH(!isUpd9002()); // R3
         resetFlagN();
         setFlagXY(*rp);
     }
@@ -4397,7 +4435,7 @@ class Z80
         setFlagZ(!n);
         setFlagPV(isFlagZ());
         setFlagS(!isFlagZ() && 7 == bit);
-        setFlagH();
+        setFlagH(!isUpd9002()); // R3
         resetFlagN();
         setFlagXY((reg.WZ & 0xFF00) >> 8);
     }
@@ -4421,7 +4459,7 @@ class Z80
         setFlagZ(!n);
         setFlagPV(isFlagZ());
         setFlagS(!isFlagZ() && 7 == bit);
-        setFlagH();
+        setFlagH(!isUpd9002()); // R3
         resetFlagN();
         setFlagXY((reg.WZ & 0xFF00) >> 8);
     }
@@ -4445,7 +4483,7 @@ class Z80
         setFlagZ(!n);
         setFlagPV(isFlagZ());
         setFlagS(!isFlagZ() && 7 == bit);
-        setFlagH();
+        setFlagH(!isUpd9002()); // R3
         resetFlagN();
         setFlagXY((reg.WZ & 0xFF00) >> 8);
     }
@@ -6170,6 +6208,7 @@ class Z80
         reg.pair.A = 0xff;
         reg.pair.F = 0xff;
         reg.SP = 0xffff;
+        maskUndocumentedFlags(); // R1
         memset(&wtc, 0, sizeof(wtc));
     }
 
-- 
2.43.0

